In [1]:
import graphlab

# Load CIFAR-10 dataset

In [2]:
img_train = graphlab.SFrame('https://d396qusza40orc.cloudfront.net/phoenixassets/image_train_data.csv')

[INFO] graphlab.cython.cy_server: GraphLab Create v2.1 started. Logging: C:\Users\ballen\AppData\Local\Temp\graphlab_server_1477663286.log.0


This non-commercial license of GraphLab Create for academic use is assigned to somewhatalive@gmail.com and will expire on October 26, 2017.


Downloading https://d396qusza40orc.cloudfront.net/phoenixassets/image_train_data.csv to C:/Users/ballen/AppData/Local/Temp/graphlab-ballen/5008/758c9a00-16e1-4246-8015-a21b1eccb665.csv

Finished parsing file https://d396qusza40orc.cloudfront.net/phoenixassets/image_train_data.csv

Parsing completed. Parsed 100 lines in 0.812604 secs.

------------------------------------------------------
Inferred types from first 100 line(s) of file as 
column_type_hints=[long,str,str,array,array]
If parsing fails due to incorrect types, you can correct
the inferred type list above and pass it to read_csv in
the column_type_hints argument
------------------------------------------------------


Read 1943 lines. Lines per second: 1646.4

Finished parsing file https://d396qusza40orc.cloudfront.net/phoenixassets/image_train_data.csv

Parsing completed. Parsed 2005 lines in 1.20015 secs.

In [3]:
img_train.head()

Columns:
	id	int
	image	str
	label	str
	deep_features	array
	image_array	array

Rows: 10

Data:
+-----+----------------------+------------+-------------------------------+
|  id |        image         |   label    |         deep_features         |
+-----+----------------------+------------+-------------------------------+
|  24 | Height: 32 Width: 32 |    bird    | [0.242872, 1.09545, 0.0, 0... |
|  33 | Height: 32 Width: 32 |    cat     | [0.525088, 0.0, 0.0, 0.0, ... |
|  36 | Height: 32 Width: 32 |    cat     | [0.566016, 0.0, 0.0, 0.0, ... |
|  70 | Height: 32 Width: 32 |    dog     | [1.1298, 0.0, 0.0, 0.77819... |
|  90 | Height: 32 Width: 32 |    bird    | [1.71787, 0.0, 0.0, 0.0, 0... |
|  97 | Height: 32 Width: 32 | automobile | [1.57819, 0.0, 0.0, 0.0, 0... |
| 107 | Height: 32 Width: 32 |    dog     | [0.0, 0.0, 0.220678, 0.0, ... |
| 121 | Height: 32 Width: 32 |    bird    | [0.0, 0.237535, 0.0, 0.0, ... |
| 136 | Height: 32 Width: 32 | automobile | [0.0, 0.0, 0.0, 0.0, 0.0, ... |
| 138 | Height: 32 Width: 32 |    bird    | [0.658936, 0.0, 0.0, 0.0, ... |
+-----+----------------------+------------+-------------------------------+
+-------------------------------+
|          image_array          |
+-------------------------------+
| [73.0, 77.0, 58.0, 71.0, 6... |
| [7.0, 5.0, 8.0, 7.0, 5.0, ... |
| [169.0, 122.0, 65.0, 131.0... |
| [154.0, 179.0, 152.0, 159.... |
| [216.0, 195.0, 180.0, 201.... |
| [33.0, 44.0, 27.0, 29.0, 4... |
| [97.0, 51.0, 31.0, 104.0, ... |
| [93.0, 96.0, 88.0, 102.0, ... |
| [35.0, 59.0, 53.0, 36.0, 5... |
| [205.0, 193.0, 195.0, 200.... |
+-------------------------------+
[10 rows x 5 columns]

# Train Nearest Neighbors model for retrieving images useing deep features

In [4]:
knn_model = graphlab.nearest_neighbors.create(img_train,
                                              features=['deep_features'],
                                              label='id')

Starting brute force nearest neighbors model training.

# Using kNN model to find similar images

In [5]:
graphlab.canvas.set_target('ipynb')
cat = img_train[18:19]
cat

Columns:
	id	int
	image	str
	label	str
	deep_features	array
	image_array	array

Rows: 1

Data:
+-----+----------------------+-------+-------------------------------+
|  id |        image         | label |         deep_features         |
+-----+----------------------+-------+-------------------------------+
| 384 | Height: 32 Width: 32 |  cat  | [1.04404, 0.0, 0.0, 0.0, 0... |
+-----+----------------------+-------+-------------------------------+
+-------------------------------+
|          image_array          |
+-------------------------------+
| [46.0, 45.0, 50.0, 47.0, 4... |
+-------------------------------+
[1 rows x 5 columns]

In [6]:
knn_model.query(cat)

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.0498753   | 21.002ms     |

| Done         |         | 100         | 251.031ms    |

+--------------+---------+-------------+--------------+

Columns:
	query_label	int
	reference_label	int
	distance	float
	rank	int

Rows: 5

Data:
+-------------+-----------------+---------------+------+
| query_label | reference_label |    distance   | rank |
+-------------+-----------------+---------------+------+
|      0      |       384       |      0.0      |  1   |
|      0      |       6910      |  36.940312418 |  2   |
|      0      |      39777      | 38.4634892031 |  3   |
|      0      |      36870      | 39.7559665724 |  4   |
|      0      |      41734      | 39.7865972971 |  5   |
+-------------+-----------------+---------------+------+
[5 rows x 4 columns]

In [7]:
def get_images_from_ids(query_result):
    return img_train.filter_by(query_result['reference_label'],'id')

In [8]:
cat_neighbors = get_images_from_ids(knn_model.query(cat))

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.0498753   | 27.503ms     |

| Done         |         | 100         | 268.534ms    |

+--------------+---------+-------------+--------------+

In [9]:
cat_neighbors['image'].show()

## No images :(

In [10]:
car = img_train[8:9]
car['image'].show()

In [11]:
car_neighbors = get_images_from_ids(knn_model.query(car))

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.0498753   | 16.502ms     |

| Done         |         | 100         | 282.536ms    |

+--------------+---------+-------------+--------------+

In [12]:
car_neighbors['image'].show()

# Just for fun, create a lambda to find and show nearest neighbor images

In [13]:
show_neighbors = lambda i: get_images_from_ids(knn_model.query(img_train[i:i+1]))['image'].show()

In [14]:
show_neighbors(8)

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.0498753   | 22.003ms     |

| Done         |         | 100         | 252.032ms    |

+--------------+---------+-------------+--------------+

# 1. Compute sketch summary statistics of data

In [15]:
img_train['label'].sketch_summary()


+------------------+-------+----------+
|       item       | value | is exact |
+------------------+-------+----------+
|      Length      |  2005 |   Yes    |
| # Missing Values |   0   |   Yes    |
| # unique values  |   4   |    No    |
+------------------+-------+----------+

Most frequent items:
+-------+------------+-----+-----+------+
| value | automobile | cat | dog | bird |
+-------+------------+-----+-----+------+
| count |    509     | 509 | 509 | 478  |
+-------+------------+-----+-----+------+


# 2. Creating category-specific image retrieval models

In [19]:
image_train = img_train

In [20]:
dog_train = image_train[image_train['label'].contains('dog')]
cat_train = image_train[image_train['label'].contains('cat')]
auto_train = image_train[image_train['label'].contains('automobile')]
bird_train = image_train[image_train['label'].contains('bird')]

### Individualized kNN models for each category

In [60]:
dog_model = graphlab.nearest_neighbors.create(dog_train,features=['deep_features'],
                                              label='id')

Starting brute force nearest neighbors model training.

In [61]:
cat_model = graphlab.nearest_neighbors.create(cat_train,features=['deep_features'],
                                              label='id')

Starting brute force nearest neighbors model training.

In [62]:
auto_model = graphlab.nearest_neighbors.create(auto_train,features=['deep_features'],
                                              label='id')

Starting brute force nearest neighbors model training.

In [63]:
bird_model = graphlab.nearest_neighbors.create(bird_train,features=['deep_features'],
                                              label='id')

Starting brute force nearest neighbors model training.

### Load test_data

In [25]:
image_test = graphlab.SFrame('https://d396qusza40orc.cloudfront.net/phoenixassets/image_test_data.csv')

Downloading https://d396qusza40orc.cloudfront.net/phoenixassets/image_test_data.csv to C:/Users/ballen/AppData/Local/Temp/graphlab-ballen/5008/2077d4e1-037b-43f0-940c-24a17da36c12.csv

Finished parsing file https://d396qusza40orc.cloudfront.net/phoenixassets/image_test_data.csv

Parsing completed. Parsed 100 lines in 0.824105 secs.

------------------------------------------------------
Inferred types from first 100 line(s) of file as 
column_type_hints=[long,str,str,array,array]
If parsing fails due to incorrect types, you can correct
the inferred type list above and pass it to read_csv in
the column_type_hints argument
------------------------------------------------------


Read 1940 lines. Lines per second: 1377.18

Finished parsing file https://d396qusza40orc.cloudfront.net/phoenixassets/image_test_data.csv

Parsing completed. Parsed 4000 lines in 2.23428 secs.

In [26]:
image_test[0:1]

Columns:
	id	int
	image	str
	label	str
	deep_features	array
	image_array	array

Rows: 1

Data:
+----+----------------------+-------+-------------------------------+
| id |        image         | label |         deep_features         |
+----+----------------------+-------+-------------------------------+
| 0  | Height: 32 Width: 32 |  cat  | [1.13469, 0.0, 0.0, 0.0, 0... |
+----+----------------------+-------+-------------------------------+
+-------------------------------+
|          image_array          |
+-------------------------------+
| [158.0, 112.0, 49.0, 159.0... |
+-------------------------------+
[1 rows x 5 columns]

In [27]:
## Querying nearest neighbor using cat-model

cat_model.query(image_test[0:1])

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.196464    | 18.002ms     |

| Done         |         | 100         | 96.012ms     |

+--------------+---------+-------------+--------------+

Columns:
	query_label	int
	reference_label	int
	distance	float
	rank	int

Rows: 5

Data:
+-------------+-----------------+---------------+------+
| query_label | reference_label |    distance   | rank |
+-------------+-----------------+---------------+------+
|      0      |      16289      | 34.6237223947 |  1   |
|      0      |      45646      | 36.0068799499 |  2   |
|      0      |      32139      | 36.5200809583 |  3   |
|      0      |      25713      | 36.7548479811 |  4   |
|      0      |       331       | 36.8731153276 |  5   |
+-------------+-----------------+---------------+------+
[5 rows x 4 columns]

In [28]:
## Querying nearest negihbor using dog model

dog_model.query(image_test[0:1])

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.209205    | 10.502ms     |

| Done         |         | 100         | 98.013ms     |

+--------------+---------+-------------+--------------+

Columns:
	query_label	int
	reference_label	int
	distance	float
	rank	int

Rows: 5

Data:
+-------------+-----------------+---------------+------+
| query_label | reference_label |    distance   | rank |
+-------------+-----------------+---------------+------+
|      0      |      49573      | 38.0742668329 |  1   |
|      0      |      47248      | 39.1015267718 |  2   |
|      0      |      16041      | 39.6991053472 |  3   |
|      0      |       4549      | 39.7039797387 |  4   |
|      0      |        90       | 39.8908818008 |  5   |
+-------------+-----------------+---------------+------+
[5 rows x 4 columns]

# 3. A simple example of nearest-neighbors classification: 


In [29]:
## Cat model mean distance

cat_model.query(image_test[0:1])['distance'].mean()

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.196464    | 18.503ms     |

| Done         |         | 100         | 91.512ms     |

+--------------+---------+-------------+--------------+

36.15572932231883

In [30]:
## Dog model mean distance

dog_model.query(image_test[0:1])['distance'].mean()

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.209205    | 17.503ms     |

| Done         |         | 100         | 108.514ms    |

+--------------+---------+-------------+--------------+

39.29395209827645

Mean distance closer for image_test[0:1] in cat_model when compared to dog_model

# 4. Computing nearest neighbors accuracy using SFrame operations:

In [64]:
## Splitting test data by label

dog_test = image_test[image_test['label'].contains('dog')]
cat_test = image_test[image_test['label'].contains('cat')]
auto_test = image_test[image_test['label'].contains('automobile')]
bird_test = image_test[image_test['label'].contains('bird')]

In [65]:
## Training inidividualized kNN models

dog_cat = cat_model.query(dog_test, k=1)
dog_dog = dog_model.query(dog_test, k=1)
dog_auto = auto_model.query(dog_test, k=1)
dog_bird = bird_model.query(dog_test, k=1)

Starting blockwise querying.

max rows per data block: 4348

number of reference data blocks: 4

number of query data blocks: 1

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 1000         | 127000  | 24.9509     | 538.068ms    |

| Done         | 509000  | 100         | 584.074ms    |

+--------------+---------+-------------+--------------+

Starting blockwise querying.

max rows per data block: 4348

number of reference data blocks: 4

number of query data blocks: 1

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 1000         | 128000  | 25.1473     | 553.57ms     |

| Done         | 509000  | 100         | 576.573ms    |

+--------------+---------+-------------+--------------+

Starting blockwise querying.

max rows per data block: 4348

number of reference data blocks: 4

number of query data blocks: 1

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 1000         | 127000  | 24.9509     | 540.568ms    |

| Done         | 509000  | 100         | 558.571ms    |

+--------------+---------+-------------+--------------+

Starting blockwise querying.

max rows per data block: 4348

number of reference data blocks: 4

number of query data blocks: 1

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 1000         | 119000  | 24.8954     | 560.571ms    |

| Done         | 478000  | 100         | 711.59ms     |

+--------------+---------+-------------+--------------+

In [66]:
dog_distance = graphlab.SFrame({'dog-dog': dog_dog['distance'],'dog-cat': dog_cat['distance'],
                               'dog-auto': dog_auto['distance'], 'dog-bird': dog_bird['distance']})

In [42]:
categorical_train_data = [dog_train, cat_train, auto_train, bird_train]
categorical_test_data = [dog_test, cat_test, auto_test, bird_test]
categorical_models = [dog_model, cat_model, auto_model, bird_model]

In [67]:
dog_distance

Columns:
	dog-auto	float
	dog-bird	float
	dog-cat	float
	dog-dog	float

Rows: 1000

Data:
+---------------+---------------+---------------+---------------+
|    dog-auto   |    dog-bird   |    dog-cat    |    dog-dog    |
+---------------+---------------+---------------+---------------+
| 41.9579817943 | 41.7538626572 | 36.4196034322 | 33.4773536247 |
| 46.0021346539 | 41.3382923691 | 38.8353256562 | 32.8458494483 |
|  42.946230695 |  38.615760847 |  36.976339279 | 35.0397069097 |
| 41.6866044216 | 37.0892224529 | 34.5750053045 | 33.9010316493 |
|  39.226972617 |  38.272290586 | 34.7788271508 | 37.4849284982 |
| 40.5845144886 | 39.1462124795 | 35.1171583313 | 34.9451677858 |
|  45.106733015 | 40.5230430151 | 40.6095824993 | 39.0957282483 |
| 41.3221134071 | 38.1947935831 | 39.9036862952 | 37.7696112376 |
| 41.8244738585 | 40.1567116178 | 38.0674687988 | 35.1089164503 |
| 45.4976975148 | 45.5597961334 | 42.7258755717 | 43.2422865878 |
+---------------+---------------+---------------+---------------+
[1000 rows x 4 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [68]:
## Selecting specfic distance within dog_disctance SFrame

dog_distance[0]['dog-cat']

36.41960343224111

In [69]:
## Helper function to check if dog-dog dostance is the lowest of each row

def is_dog_correct(row):
    if row['dog-dog'] < row['dog-cat'] and row['dog-dog'] < row['dog-bird'] and row['dog-dog'] < row['dog-auto']:
        return 1
    else:
        return 0  

In [70]:
## Checking to see if function works, it does!

is_dog_correct(dog_distance[1])

1

In [73]:
## Finding accuracy of our dog-dog model

dog_distance.apply(is_dog_correct).sum()

678L

Out of 1000 dog images, 678 are closer to our dog kernal than all other kernels based on our categorical kNN model